File: leetcode/algorithms/139_word_break/main.py
from typing import List, Set


class WordBreak:
    """
    # Dynamic Programming
    #
    # Complexities:
    #   N - Size of `s`
    #   - Time Complexity: O(N³)
    #   - Space Complexity: O(N)
    """
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        s_len: int = len(s)

        dict: Set[str] = set(wordDict)

        dp: List[bool] = [False] * (s_len + 1)
        dp[0] = True

        for i in range(1, s_len + 1):
            for j in range(0, i):
                if dp[j] and s[j:i] in dict:
                    dp[i] = True
                    break

        return dp[s_len]


    # Solution
    """
    # Dynamic Programming
    #
    # Complexities:
    #   N - Size of `s`
    #   - Time Complexity: O(N³)
    #   - Space Complexity: O(N)
    """
    def solution(self, s: str, wordDict: List[str]) -> bool:
        word_set = set(wordDict)
        n = len(s)

        dp = [False] * (n + 1)
        dp[0] = True

        for i in range(1, n + 1):
            for j in range(i):
                if dp[j] and s[j:i] in word_set:
                    dp[i] = True
                    break

        return dp[n]

File: leetcode/algorithms/139_word_break/main.dart
class WordBreak {
  /**
   * Dynamic Programming
   *
   * Complexities:
   *   N - Size of `s`
   *   - Time Complexity: O(N³)
   *   - Space Complexity: O(N)
   */
  bool wordBreak(String s, List<String> wordDict) {
    final int sLen = s.length;

    final Set<String> dict = Set.from(wordDict);

    final List<bool> dp = List.filled(sLen + 1, false);
    dp[0] = true;

    for (int i = 1; i <= sLen; i++) {
      for (int j = 0; j < i; j++) {
        if (dp[j] && dict.contains(s.substring(j, i))) {
          dp[i] = true;
          break;
        }
      }
    }

    return dp[sLen];
  }


  // Solution
  /**
   * Dynamic Programming
   *
   * Complexities:
   *   N - Size of `s`
   *   - Time Complexity: O(N³)
   *   - Space Complexity: O(N)
   */
  bool solution(String s, List<String> wordDict) {
    final Set<String> wordSet = wordDict.toSet();

    final int n = s.length;
    List<bool> dp = List.filled(n + 1, false);

    dp[0] = true;

    for (int i = 1; i <= n; i++) {
      for (int j = 0; j < i; j++) {
        if (dp[j] && wordSet.contains(s.substring(j, i))) {
          dp[i] = true;
          break;
        }
      }
    }

    return dp[n];
  }
}
